File: ValonSynth.cc
//# Copyright (C) 2011 Associated Universities, Inc. Washington DC, USA.
//# 
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//# 
//# This program is distributed in the hope that it will be useful, but
//# WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
//# General Public License for more details.
//# 
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//# 
//# Correspondence concerning GBT software should be addressed as follows:
//#    GBT Operations
//#    National Radio Astronomy Observatory
//#    P. O. Box 2
//#    Green Bank, WV 24944-0002 USA

#include "Serial.h"
#include "ValonSynth.h"


ValonSynth::ValonSynth(const char *port)
    :
    s(port)
{
}

//------------------//
// Output Frequency //
//------------------//
bool
ValonSynth::get_frequency(enum ValonSynth::Synthesizer synth, float &frequency)
{
    std::uint32_t ncount, frac, mod, dbf;
    bool success = read_freq_registers(synth, ncount, frac, mod, dbf);
    float EPDF = getEPDF(synth);
    frequency = (ncount + float(frac) / mod) * EPDF / dbf;
    return success;
}

bool
ValonSynth::set_frequency(enum ValonSynth::Synthesizer synth, float frequency,
                          float chan_spacing)
{
    vco_range vcor;
    std::uint32_t dbf = 1;
    bool success = get_vco_range(synth, vcor);
    while(((frequency * dbf) <= vcor.min) && (dbf <= 16))
    {
        dbf *= 2;
    }
    if(dbf > 16)
    {
        dbf = 16;
    }
    float vco = frequency * dbf;
    float EPDF = getEPDF(synth);
    std::uint32_t ncount = std::uint32_t(vco / EPDF);
    std::uint32_t frac = std::uint32_t((vco - ncount * EPDF) / chan_spacing + 0.5);
    std::uint32_t mod = std::uint32_t(EPDF / chan_spacing + 0.5);
    // Reduce frac/mod to simplest fraction
    if((frac != 0) && (mod != 0))
    {
        while(!(frac & 1) && !(mod & 1))
        {
            frac /= 2;
            mod /= 2;
        }
    }
    else
    {
        frac = 0;
        mod = 1;
    }
    return success && write_freq_registers(synth, ncount, frac, mod, dbf);
}

//---------------------//
// Reference Frequency //
//---------------------//
bool
ValonSynth::get_reference(std::uint32_t &frequency)
{
    std::uint8_t bytes[4];
    std::uint8_t checksum;
    bytes[0] = 0x81;
    s.write(bytes, 1);
    s.read(bytes, 4);
    s.read(&checksum, 1);
#ifdef VERIFY_CHECKSUM
    if (!verify_checksum(bytes, 4, checksum)) return false;
#endif//VERIFY_CHECKSUM
    unpack_int(bytes, frequency);
    return true;
}

bool
ValonSynth::set_reference(std::uint32_t frequency)
{
    std::uint8_t bytes[6];
    bytes[0] = 0x01;
    pack_int(frequency, &bytes[1]);
    bytes[5] = generate_checksum(bytes, 5);
    s.write(bytes, 6);
    s.read(bytes, 1);
    return bytes[0] == ACK;
}

//----------//
// RF Level //
//----------//
bool
ValonSynth::get_rf_level(enum ValonSynth::Synthesizer synth, std::int32_t &rf_level)
{
    reg0_t r0;
    reg1_t r1;
    reg2_t r2;
    reg3_t r3;
    reg4_t r4;
    reg5_t r5;
    bool success = get_all_registers(synth, r0, r1, r2, r3, r4, r5);
    switch(static_cast<std::uint32_t>(r4.output_power))
    {
    case 0: rf_level = -4; break;
    case 1: rf_level = -1; break;
    case 2: rf_level =  2; break;
    case 3: rf_level =  5; break;
    }
    return success;
}

bool
ValonSynth::set_rf_level(enum ValonSynth::Synthesizer synth, int32_t rf_level)
{
    std::uint32_t rfl;
    switch(rf_level)
    {
    case -4: rfl = 0; break;
    case -1: rfl = 1; break;
    case 2:  rfl = 2; break;
    case 5:  rfl = 3; break;
    default: return false;
    }
    reg0_t r0;
    reg1_t r1;
    reg2_t r2;
    reg3_t r3;
    reg4_t r4;
    reg5_t r5;
    bool success = get_all_registers(synth, r0, r1, r2, r3, r4, r5);
    r4.output_power = rfl;
    return success && set_all_registers(synth, r0, r1, r2, r3, r4, r5);
}

//---------------------//
// ValonSynth Options //
//---------------------//
bool
ValonSynth::get_options(enum ValonSynth::Synthesizer synth, options &opts)
{
    reg0_t r0;
    reg1_t r1;
    reg2_t r2;
    reg3_t r3;
    reg4_t r4;
    reg5_t r5;
    bool success = get_all_registers(synth, r0, r1, r2, r3, r4, r5);
    opts.double_ref = r2.double_r;
    opts.half_ref   = r2.half_r;
    opts.r          = r2.r;
    opts.low_spur   = bool(r2.low_spur);
    return success;
}

bool
ValonSynth::set_options(enum ValonSynth::Synthesizer synth,
                        const options &opts)
{
    reg0_t r0;
    reg1_t r1;
    reg2_t r2;
    reg3_t r3;
    reg4_t r4;
    reg5_t r5;
    bool success = get_all_registers(synth, r0, r1, r2, r3, r4, r5);
    r2.double_r = opts.double_ref;
    r2.half_r   = opts.half_ref;
    r2.r        = opts.r;
    // Value must be either 0x0 or 0x3
    r2.low_spur = opts.low_spur ? 3 : 0;
    return success && set_all_registers(synth, r0, r1, r2, r3, r4, r5);
}

//------------------//
// Reference Select //
//------------------//
bool
ValonSynth::get_ref_select(bool &e_not_i)
{
    std::uint8_t bytes;
    std::uint8_t checksum;
    bytes = 0x86;
    s.write(&bytes, 1);
    s.read(&bytes, 1);
    s.read(&checksum, 1);
#ifdef VERIFY_CHECKSUM
    if(!verify_checksum(&bytes, 1, checksum)) return false;
#endif//VERIFY_CHECKSUM
    e_not_i = bytes & 1;
    return true;
}

bool
ValonSynth::set_ref_select(bool e_not_i)
{
    std::uint8_t bytes[3];
    bytes[0] = 0x06;
    bytes[1] = e_not_i & 1;
    bytes[2] = generate_checksum(bytes, 2);
    s.write(bytes, 3);
    s.read(bytes, 1);
    return bytes[0] == ACK;
}

//-----------//
// VCO Range //
//-----------//
bool
ValonSynth::get_vco_range(enum ValonSynth::Synthesizer synth, vco_range &vcor)
{
    std::uint8_t bytes[4];
    std::uint8_t checksum;
    bytes[0] = 0x83 | synth;
    s.write(bytes, 1);
    s.read(bytes, 4);
    s.read(&checksum, 1);
#ifdef VERIFY_CHECKSUM
    if(!verify_checksum(bytes, 4, checksum)) return false;
#endif//VERIFY_CHECKSUM
    unpack_short(&bytes[0], vcor.min);
    unpack_short(&bytes[2], vcor.max);
    return true;
}

bool
ValonSynth::set_vco_range(enum ValonSynth::Synthesizer synth,
                          const vco_range &vcor)
{
    std::uint8_t bytes[6];
    bytes[0] = 0x03 | synth;
    pack_short(vcor.min, &bytes[1]);
    pack_short(vcor.max, &bytes[3]);
    bytes[5] = generate_checksum(bytes, 5);
    s.write(bytes, 6);
    s.read(bytes, 1);
    return bytes[0] == ACK;
}

//------------//
// Phase Lock //
//------------//
bool
ValonSynth::get_phase_lock(enum ValonSynth::Synthesizer synth, bool &locked)
{
    std::uint8_t bytes;
    std::uint8_t checksum;
    bytes = 0x86 | synth;
    s.write(&bytes, 1);
    s.read(&bytes, 1);
    s.read(&checksum, 1);
#ifdef VERIFY_CHECKSUM
    if(!verify_checksum(&bytes, 1, checksum)) return false;
#endif//VERIFY_CHECKSUM
    std::uint32_t mask;
    // ValonSynth A
    if(synth == ValonSynth::A) mask = 0x20;
    // ValonSynth B
    else mask = 0x10;
    locked = bytes & mask;
    return true;
}

//-------------------//
// ValonSynth Label //
//-------------------//
bool
ValonSynth::get_label(enum ValonSynth::Synthesizer synth, char *label)
{
    std::uint8_t bytes[16];
    std::uint8_t checksum;
    bytes[0] = 0x82 | synth;
    s.write(bytes, 1);
    s.read(bytes, 16);
    s.read(&checksum, 1);
#ifdef VERIFY_CHECKSUM
    if(!verify_checksum(bytes, 16, checksum)) return false;
#endif//VERIFY_CHECKSUM
    memcpy(label, (char*)bytes, 16);
    return true;
}

bool
ValonSynth::set_label(enum ValonSynth::Synthesizer synth, const char *label)
{
    std::uint8_t bytes[18];
    bytes[0] = 0x02 | synth;
    memcpy(&bytes[1], label, 16);
    bytes[17] = generate_checksum(bytes, 17);
    s.write(bytes, 18);
    s.read(bytes, 1);
    return bytes[0] == ACK;
}

//-------//
// Flash //
//-------//
bool
ValonSynth::flash()
{
    std::uint8_t bytes[2];
    bytes[0] = 0x40;
    bytes[1] = generate_checksum(bytes, 1);
    s.write(bytes, 2);
    s.read(bytes, 1);
    return bytes[0] == ACK;
}

//------------------//
// EDPF Calculation //
//------------------//
float
ValonSynth::getEPDF(enum ValonSynth::Synthesizer synth)
{
    options opts;
    float reference = get_reference() / 1e6;
    get_options(synth, opts);

    if(opts.double_ref) reference *= 2.0;
    if(opts.half_ref) reference /= 2.0;
    if(opts.r > 1) reference /= opts.r;
    return reference;
}

//----------//
// Checksum //
//----------//
std::uint8_t
ValonSynth::generate_checksum(const std::uint8_t *bytes, std::size_t length)
{
    std::uint32_t sum = 0;
    for(std::size_t i = 0; i < length; ++i)
    {
        sum += bytes[i];
    }
    return (uint8_t)(sum % 256);
}

bool
ValonSynth::verify_checksum(const std::uint8_t *bytes, std::size_t length,
                            std::uint8_t checksum)
{
    return (generate_checksum(bytes, length) == checksum);
}

//-------------//
// Bit Packing //
//-------------//
bool ValonSynth::write_freq_registers(enum ValonSynth::Synthesizer synth,
                                      std::uint32_t ncount, std::uint32_t frac,
                                      std::uint32_t mod, std::uint32_t dbf) {
    std::uint32_t raw_dbf;
    switch(dbf)
    {
    case  1: raw_dbf = 0; break;
    case  2: raw_dbf = 1; break;
    case  4: raw_dbf = 2; break;
    case  8: raw_dbf = 3; break;
    case 16: raw_dbf = 4; break;
    }

    reg0_t r0;
    reg1_t r1;
    reg2_t r2;
    reg3_t r3;
    reg4_t r4;
    reg5_t r5;

    bool success = get_all_registers(synth, r0, r1, r2, r3, r4, r5);

    r0.ncount = ncount;
    r0.frac = frac;
    r1.mod = mod;
    r4.divider_select = raw_dbf;

    return success && set_all_registers(synth, r0, r1, r2, r3, r4, r5);
}

bool ValonSynth::read_freq_registers(enum ValonSynth::Synthesizer synth,
                                     std::uint32_t &ncount, std::uint32_t &frac,
                                     std::uint32_t &mod, std::uint32_t &dbf) {
    reg0_t r0;
    reg1_t r1;
    reg2_t r2;
    reg3_t r3;
    reg4_t r4;
    reg5_t r5;

    bool success = get_all_registers(synth, r0, r1, r2, r3, r4, r5);

    ncount = r0.ncount;
    frac = r0.frac;
    mod = r1.mod;
    switch(static_cast<std::uint32_t>(r4.divider_select))
    {
    case  0: dbf = 1; break;
    case  1: dbf = 2; break;
    case  2: dbf = 4; break;
    case  3: dbf = 8; break;
    case  4: dbf = 16; break;
    default: dbf = 1;
    }

    return success;
}

bool ValonSynth::get_all_registers(enum ValonSynth::Synthesizer synth,
                       reg0_t &r0, reg1_t &r1,
                       reg2_t &r2, reg3_t &r3,
                       reg4_t &r4, reg5_t &r5) {
    std::uint8_t bytes[24];
    std::uint8_t checksum;
    bytes[0] = 0x80 | synth;
    s.write(bytes, 1);
    s.read(bytes, 24);
    s.read(&checksum, 1);
#ifdef VERIFY_CHECKSUM
    if(!verify_checksum(bytes, 24, checksum)) return false;
#endif//VERIFY_CHECKSUM
    unpack_int(&bytes[ 0], r0.asbyte);
    unpack_int(&bytes[ 4], r1.asbyte);
    unpack_int(&bytes[ 8], r2.asbyte);
    unpack_int(&bytes[12], r3.asbyte);
    unpack_int(&bytes[16], r4.asbyte);
    unpack_int(&bytes[20], r5.asbyte);
    return true;
}

bool ValonSynth::set_all_registers(enum ValonSynth::Synthesizer synth,
                                   const reg0_t &r0, const reg1_t &r1,
                                   const reg2_t &r2, const reg3_t &r3,
                                   const reg4_t &r4, const reg5_t &r5) {
    std::uint8_t bytes[26];
    bytes[0] = 0x00 | synth;
    pack_int(r0.asbyte, &bytes[0]);
    pack_int(r1.asbyte, &bytes[4]);
    pack_int(r2.asbyte, &bytes[8]);
    pack_int(r3.asbyte, &bytes[12]);
    pack_int(r4.asbyte, &bytes[16]);
    pack_int(r5.asbyte, &bytes[20]);
    bytes[25] = generate_checksum(bytes, 25);
    s.write(bytes, 26);
    s.read(bytes, 1);
    return bytes[0] == ACK;
}

void
ValonSynth::pack_int(std::uint32_t num, std::uint8_t *bytes)
{
    bytes[0] = (num >> 24) & 0xff;
    bytes[1] = (num >> 16) & 0xff;
    bytes[2] = (num >> 8) & 0xff;
    bytes[3] = (num) & 0xff;
}

void
ValonSynth::pack_short(std::uint16_t num, std::uint8_t *bytes)
{
    bytes[0] = (num >> 8) & 0xff;
    bytes[1] = (num) & 0xff;
}

void
ValonSynth::unpack_int(const std::uint8_t *bytes, std::uint32_t &num)
{
    num = ((std::uint32_t(bytes[0]) << 24) + (std::uint32_t(bytes[1]) << 16) +
           (std::uint32_t(bytes[2]) <<  8) + (std::uint32_t(bytes[3])));
}

void
ValonSynth::unpack_short(const std::uint8_t *bytes, std::uint16_t &num)
{
    num = (std::uint16_t(bytes[0]) << 8) + std::uint16_t(bytes[1]);
}

File: src/valon_synth.py
# Copyright (C) 2011 Associated Universities, Inc. Washington DC, USA.
# 
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
# 
# Correspondence concerning GBT software should be addressed as follows:
#	GBT Operations
#	National Radio Astronomy Observatory
#	P. O. Box 2
#	Green Bank, WV 24944-0002 USA

"""
Provides a serial interface to the Valon 500x.
"""

# Python modules
import struct
# Third party modules
import serial


__author__ = "Patrick Brandt"
__copyright__ = "Copyright 2011, Associated Universities, Inc."
__credits__ = ["Patrick Brandt, Stewart Rumley, Steven Stark"]
__license__ = "GPL"
#__version__ = "1.0"
__maintainer__ = "Patrick Brandt"


# Handy aliases
SYNTH_A = 0x00
SYNTH_B = 0x08

INT_REF = 0x00
EXT_REF = 0x01

ACK = 0x06
NACK = 0x15

def _generate_checksum(data):
    "Generate a checksum for the data provided."
    return chr(sum([ord(b) for b in data]) % 256)

def _verify_checksum(data, checksum):
    "Verify a checksum for the data provided."
    return (_generate_checksum(data) == checksum)

def _pack_freq_registers(ncount, frac, mod, dbf, old_data):
    "Do bit packing for the frequency setting registers."
    dbf_table = {1: 0, 2: 1, 4: 2, 8: 3, 16: 4}
    reg0, reg1, reg2, reg3, reg4, reg5 = struct.unpack('>IIIIII', old_data)
    reg0 &= 0x80000007
    reg0 |= ((ncount & 0xffff) << 15) | ((frac & 0x0fff) << 3)
    reg1 &= 0xffff8007
    reg1 |= (mod & 0x0fff) << 3
    reg4 &= 0xff8fffff
    reg4 |= (dbf_table.get(dbf, 0)) << 20
    return struct.pack('>IIIIII', reg0, reg1, reg2, reg3, reg4, reg5)

def _unpack_freq_registers(data):
    "Do bit unpacking for the frequency setting registers."
    dbf_rev_table = {0: 1, 1: 2, 2: 4, 3: 8, 4: 16}
    reg0, reg1, _, _, reg4, _ = struct.unpack('>IIIIII', data)
    ncount = (reg0 >> 15) & 0xffff
    frac = (reg0 >> 3) & 0x0fff
    mod = (reg1 >> 3) & 0x0fff
    dbf = dbf_rev_table.get((reg4 >> 20) & 0x07, 1)
    return ncount, frac, mod, dbf

class Synthesizer:
    """A simple interface to the Valon 500x synthesizer."""
    def __init__(self, port):
        self.conn = serial.Serial(None, 9600, serial.EIGHTBITS,
                                  serial.PARITY_NONE, serial.STOPBITS_ONE)
        self.conn.setPort(port)

    def get_frequency(self, synth):
        """
        Returns the current output frequency for the selected synthesizer.

        @param synth : synthesizer this command affects (0 for 1, 8 for 2).
        @type  synth : int

        @return: the frequency in MHz (float)
        """
        self.conn.open()
        data = struct.pack('>B', 0x80 | synth)
        self.conn.write(data)
        data = self.conn.read(24)
        checksum = self.conn.read(1)
        self.conn.close()
        #_verify_checksum(data, checksum)
        ncount, frac, mod, dbf = _unpack_freq_registers(data)
        epdf = self._get_epdf(synth)
        return (ncount + float(frac) / mod) * epdf / dbf

    def set_frequency(self, synth, freq, chan_spacing = 10.):
        """
        Sets the synthesizer to the desired frequency

        Sets to the closest possible frequency, depending on the channel spacing.
        Range is determined by the minimum and maximum VCO frequency.

        @param synth : synthesizer this command affects (0 for 1, 8 for 2).
        @type  synth : int

        @param freq : output frequency
        @type  freq : float

        @param chan_spacing : output frequency increment
        @type  chan_spacing : float

        @return: True if success (bool)
        """
        low, _ = self.get_vco_range(synth)
        dbf = 1
        while (freq * dbf) <= low and dbf <= 16:
            dbf *= 2
        if dbf > 16:
            dbf = 16
        vco = freq * dbf
        epdf = self._get_epdf(synth)
        ncount = int(vco / epdf)
        frac = int((vco - ncount * float(epdf)) / chan_spacing + 0.5)
        mod = int(epdf / float(chan_spacing) + 0.5)
        if frac != 0 and mod != 0:
            while not (frac & 1) and not (mod & 1):
                frac /= 2
                mod /= 2
        else:
            frac = 0
            mod = 1
        self.conn.open()
        data = struct.pack('>B', 0x80 | synth)
        self.conn.write(data)
        old_data = self.conn.read(24)
        checksum = self.conn.read(1)
        #_verify_checksum(old_data, checksum)
        data = struct.pack('>B24s', 0x00 | synth,
                           _pack_freq_registers(ncount, frac, mod,
                                                dbf, old_data))
        checksum = _generate_checksum(data)
        self.conn.write(data + checksum)
        data = self.conn.read(1)
        self.conn.close()
        ack = struct.unpack('>B', data)[0]
        return ack == ACK

    def get_reference(self):
        """
        Get reference frequency in MHz
        """
        self.conn.open()
        data = struct.pack('>B', 0x81)
        self.conn.write(data)
        data = self.conn.read(4)
        checksum = self.conn.read(1)
        self.conn.close()
        #_verify_checksum(data, checksum)
        freq = struct.unpack('>I', data)[0]
        return freq

    def set_reference(self, freq):
        """
        Set reference frequency in MHz

        @param freq : frequency in MHz
        @type  freq : float

        @return: True if success (bool)
        """
        self.conn.open()
        data = struct.pack('>BI', 0x01, freq)
        checksum = _generate_checksum(data)
        self.conn.write(data + checksum)
        data = self.conn.read(1)
        self.conn.close()
        ack = struct.unpack('>B', data)[0]
        return ack == ACK

    def get_rf_level(self, synth):
        """
        Returns RF level in dBm

        @param synth : synthesizer address, 0 or 8
        @type  synth : int

        @return: dBm (int)
        """
        rfl_table = {0: -4, 1: -1, 2: 2, 3: 5}
        self.conn.open()
        data = struct.pack('>B', 0x80 | synth)
        self.conn.write(data)
        data = self.conn.read(24)
        checksum = self.conn.read(1)
        self.conn.close()
        #_verify_checksum(data, checksum)
        _, _, _, _, reg4, _ = struct.unpack('>IIIIII', data)
        rfl = (reg4 >> 3) & 0x03
        rf_level = rfl_table.get(rfl)
        return rf_level

    def set_rf_level(self, synth, rf_level):
        """
        Set RF level

        @param synth : synthesizer address, 0 or 8
        @type  synth : int

        @param rf_level : RF power in dBm
        @type  rf_level : int

        @return: True if success (bool)
        """
        rfl_rev_table = {-4: 0, -1: 1, 2: 2, 5: 3}
        rfl = rfl_rev_table.get(rf_level)
        if(rfl is None):
            return False
        self.conn.open()
        data = struct.pack('>B', 0x80 | synth)
        self.conn.write(data)
        data = self.conn.read(24)
        checksum = self.conn.read(1)
        #_verify_checksum(data, checksum)
        reg0, reg1, reg2, reg3, reg4, reg5 = struct.unpack('>IIIIII', data)
        reg4 &= 0xffffffe7
        reg4 |= (rfl & 0x03) << 3
        data = struct.pack('>BIIIIII', 0x00 | synth,
                            reg0, reg1, reg2, reg3, reg4, reg5)
        checksum = _generate_checksum(data)
        self.conn.write(data + checksum)
        data = self.conn.read(1)
        self.conn.close()
        ack = struct.unpack('>B', data)[0]
        return ack == ACK

    def get_options(self, synth):
        """
        Get options tuple:

        bool double:   if True, reference frequency is doubled
        bool half:     if True, reference frequency is halved
        int  r:        reference frequency divisor
        bool low_spur: if True, minimizes PLL spurs;
                       if False, minimizes phase noise
        double and half both True is same as both False.

        @param synth : synthesizer address

        @return: double (bool), half (bool), r (int), low_spur (bool)
        """
        self.conn.open()
        data = struct.pack('>B', 0x80 | synth)
        self.conn.write(data)
        data = self.conn.read(24)
        checksum = self.conn.read(1)
        self.conn.close()
        #_verify_checksum(data, checksum)
        _, _, reg2, _, _, _ = struct.unpack('>IIIIII', data)
        low_spur = ((reg2 >> 30) & 1) & ((reg2 >> 29) & 1)
        double = (reg2 >> 25) & 1
        half = (reg2 >> 24) & 1
        divider = (reg2 >> 14) & 0x03ff
        return double, half, divider, low_spur

    def set_options(self, synth, double = 0, half = 0, divider = 1,
                    low_spur = 0):
        """
        Set options.

        double and half both True is same as both False.

        @param synth : synthesizer base address
        @type  synth : int

        @param double : if 1, reference frequency is doubled; default 0
        @type  double : int

        @param half : if 1, reference frequency is halved; default 0
        @type  half : int

        @param divider : reference frequency divisor; default 1
        @type  divider : int

        @param low_spur : if 1, minimizes PLL spurs;
                          if 0, minimizes phase noise; default 0
        @type  low_spur : int

        @return: True if success (bool)
        """
        self.conn.open()
        data = struct.pack('>B', 0x80 | synth)
        self.conn.write(data)
        data = self.conn.read(24)
        checksum = self.conn.read(1)
        #_verify_checksum(data, checksum)
        reg0, reg1, reg2, reg3, reg4, reg5 = struct.unpack('>IIIIII', data)
        reg2 &= 0x9c003fff
        reg2 |= (((low_spur & 1) << 30) | ((low_spur & 1) << 29) |
                 ((double & 1) << 25) | ((half & 1) << 24) |
                 ((divider & 0x03ff) << 14))
        data = struct.pack('>BIIIIII', 0x00 | synth,
                            reg0, reg1, reg2, reg3, reg4, reg5)
        checksum = _generate_checksum(data)
        self.conn.write(data + checksum)
        data = self.conn.read(1)
        self.conn.close()
        ack = struct.unpack('>B', data)[0]
        return ack == ACK

    def get_ref_select(self):
        """Returns the currently selected reference clock.

        Returns 1 if the external reference is selected, 0 otherwise.
        """
        self.conn.open()
        data = struct.pack('>B', 0x86)
        self.conn.write(data)
        data = self.conn.read(1)
        checksum = self.conn.read(1)
        self.conn.close()
        #_verify_checksum(data, checksum)
        is_ext = struct.unpack('>B', data)[0]
        return is_ext & 1

    def set_ref_select(self, e_not_i = 1):
        """
        Selects either internal or external reference clock.

        @param e_not_i : 1 (external) or 0 (internal); default 1
        @type  e_not_i : int

        @return: True if success (bool)
        """
        self.conn.open()
        data = struct.pack('>BB', 0x06, e_not_i & 1)
        checksum = _generate_checksum(data)
        self.conn.write(data + checksum)
        data = self.conn.read(1)
        self.conn.close()
        ack = struct.unpack('>B', data)[0]
        return ack == ACK

    def get_vco_range(self, synth):
        """
        Returns (min, max) VCO range tuple.

        @param synth : synthesizer base address
        @type  synth : int

        @return: min,max in MHz
        """
        self.conn.open()
        data = struct.pack('>B', 0x83 | synth)
        self.conn.write(data)
        data = self.conn.read(4)
        checksum = self.conn.read(1)
        self.conn.close()
        #_verify_checksum(data, checksum)
        return struct.unpack('>HH', data)

    def set_vco_range(self, synth, low, high):
        """
        Sets VCO range.

        @param synth : synthesizer base address
        @type  synth : int

        @param min : minimum VCO frequency
        @type  min : int

        @param max : maximum VCO frequency
        @type  max : int

        @return: True if success (bool)
        """
        self.conn.open()
        data = struct.pack('>BHH', 0x03 | synth, low, high)
        checksum = _generate_checksum(data)
        self.conn.write(data + checksum)
        data = self.conn.read(1)
        self.conn.close()
        ack = struct.unpack('>B', data)[0]
        return ack == ACK

    def get_phase_lock(self, synth):
        """
        Get phase lock status

        @param synth : synthesizer base address
        @type  synth : int

        @return: True if locked (bool)
        """
        self.conn.open()
        data = struct.pack('>B', 0x86 | synth)
        self.conn.write(data)
        data = self.conn.read(1)
        checksum = self.conn.read(1)
        self.conn.close()
        #_verify_checksum(data, checksum)
        mask = (synth << 1) or 0x20
        lock = struct.unpack('>B', data)[0] & mask
        return lock > 0

    def get_label(self, synth):
        """
        Get synthesizer label or name

        @param synth : synthesizer base address
        @type  synth : int

        @return: str
        """
        self.conn.open()
        data = struct.pack('>B', 0x82 | synth)
        self.conn.write(data)
        data = self.conn.read(16)
        checksum = self.conn.read(1)
        self.conn.close()
        #_verify_checksum(data, checksum)
        return data

    def set_label(self, synth, label):
        """
        Set synthesizer label or name

        @param synth : synthesizer base address
        @type  synth : int

        @param label : up to 16 data of text
        @type  label : str

        @return: True if success (bool)
        """
        self.conn.open()
        data = struct.pack('>B16s', 0x02 | synth, label)
        checksum = _generate_checksum(data)
        self.conn.write(data + checksum)
        data = self.conn.read(1)
        self.conn.close()
        ack = struct.unpack('>B', data)[0]
        return ack == ACK

    def flash(self):
        """
        Flash current settings for both synthesizers into non-volatile memory.

        @return: True if success (bool)
        """
        self.conn.open()
        data = struct.pack('>B', 0x40)
        checksum = _generate_checksum(data)
        self.conn.write(data + checksum)
        data = self.conn.read(1)
        self.conn.close()
        ack = struct.unpack('>B', data)[0]
        return ack == ACK

    def _get_epdf(self, synth):
        """
        Returns effective phase detector frequency.

        This is the reference frequency with options applied.
        """
        reference = self.get_reference() / 1e6
        double, half, divider, _ = self.get_options(synth)
        if(double):
            reference *= 2.0
        if(half):
            reference /= 2.0
        if(divider > 1):
            reference /= divider
        return reference
